milli/src/search/criteria/asc_desc.rs

```diff
@@ -196,30 +197,63 @@ fn facet_ordered_iterative<'t>(
 }
 
+fn facet_extreme_value<'t>(
+    mut extreme_it: impl Iterator<Item = heed::Result<(RoaringBitmap, &'t [u8])>> + 't,
+) -> Result<Option<f64>> {
+    let extreme_value =
+        if let Some(extreme_value) = extreme_it.next() { extreme_value } else { return Ok(None) };
+    let (_, extreme_value) = extreme_value?;
+
+    Ok(OrderedF64Codec::bytes_decode(extreme_value))
+}
+
+pub fn facet_min_value<'t>(
+    index: &'t Index,
+    rtxn: &'t heed::RoTxn,
+    field_id: FieldId,
+    candidates: RoaringBitmap,
+) -> Result<Option<f64>> {
+    let db = index.facet_id_f64_docids.remap_key_type::<FacetGroupKeyCodec<ByteSliceRefCodec>>();
+    let it = ascending_facet_sort(rtxn, db, field_id, candidates)?;
+    facet_extreme_value(it)
+}
+
+pub fn facet_max_value<'t>(
+    index: &'t Index,
+    rtxn: &'t heed::RoTxn,
+    field_id: FieldId,
+    candidates: RoaringBitmap,
+) -> Result<Option<f64>> {
+    let db = index.facet_id_f64_docids.remap_key_type::<FacetGroupKeyCodec<ByteSliceRefCodec>>();
+    let it = descending_facet_sort(rtxn, db, field_id, candidates)?;
+    facet_extreme_value(it)
+}
+
 fn facet_ordered_set_based<'t>(
     index: &'t Index,
     rtxn: &'t heed::RoTxn,
     field_id: FieldId,
     is_ascending: bool,
     candidates: RoaringBitmap,
 ) -> Result<Box<dyn Iterator<Item = heed::Result<RoaringBitmap>> + 't>> {
-    let make_iter = if is_ascending { ascending_facet_sort } else { descending_facet_sort };
+    let number_db =
+        index.facet_id_f64_docids.remap_key_type::<FacetGroupKeyCodec<ByteSliceRefCodec>>();
+    let string_db =
+        index.facet_id_string_docids.remap_key_type::<FacetGroupKeyCodec<ByteSliceRefCodec>>();
 
-    let number_iter = make_iter(
-        rtxn,
-        index.facet_id_f64_docids.remap_key_type::<FacetGroupKeyCodec<ByteSliceRefCodec>>(),
-        field_id,
-        candidates.clone(),
-    )?;
+    let (number_iter, string_iter) = if is_ascending {
+        let number_iter = ascending_facet_sort(rtxn, number_db, field_id, candidates.clone())?;
+        let string_iter = ascending_facet_sort(rtxn, string_db, field_id, candidates)?;
 
-    let string_iter = make_iter(
-        rtxn,
-        index.facet_id_string_docids.remap_key_type::<FacetGroupKeyCodec<ByteSliceRefCodec>>(),
-        field_id,
-        candidates,
-    )?;
+        (itertools::Either::Left(number_iter), itertools::Either::Left(string_iter))
+    } else {
+        let number_iter = descending_facet_sort(rtxn, number_db, field_id, candidates.clone())?;
+        let string_iter = descending_facet_sort(rtxn, string_db, field_id, candidates)?;
+
+        (itertools::Either::Right(number_iter), itertools::Either::Right(string_iter))
+    };
 
-    Ok(Box::new(number_iter.chain(string_iter)))
+    Ok(Box::new(number_iter.chain(string_iter).map(|res| res.map(|(doc_ids, _)| doc_ids))))
 }
 
 /// Returns an iterator over groups of the given candidates in ascending or descending order.
```

Comment:
👍 
However, I don't really understand why we use an `impl Iterator` instead of `Box<dyn Iterator` now. Was that necessary in order to put a pointer to the DB key inside?

Reply:
I'm honestly not sure anymore if it was necesssary, I believe I just changed it as part of my refactoring. I like to avoid allocations when practical, but that line is not always easy to find. I don't mind changing it back if you prefer.